File: go/regex/regex_test.go
package regex

import (
	"github.com/stretchr/testify/assert"
	"math/rand/v2"
	"regexp"
	"testing"
)

func TestRandomCases(t *testing.T) {
	var random = rand.New(rand.NewPCG(0, 0))

	for test := 0; test < 1_000_000; test++ {
		str := make([]byte, random.IntN(50))
		strFreq := 1 + random.IntN(20)
		for i := range str {
			str[i] = byte('a' + random.IntN(strFreq))
		}

		pat := make([]byte, 1+random.IntN(20))
		dotProb := 1 + random.IntN(20)
		patFreq := 1 + random.IntN(20)
		for j := 0; j < len(pat); j++ {
			pat[j] = byte('a' + random.IntN(patFreq))
			if random.IntN(100) < dotProb {
				pat[j] = '.'
			}

			if j+1 < len(pat) {
				switch prob := random.IntN(100); {
				case prob < 20:
					j++
					pat[j] = '?'
				case prob < 40:
					j++
					pat[j] = '+'
				case prob < 60:
					j++
					pat[j] = '*'
				}
			}
		}

		want, err := regexp.Match("^"+string(pat)+"$", str)
		assert.NoError(t, err)

		got := matchesFully(string(str), string(pat))
		if got != want {
			t.Fatalf("matchesFully(%q, %q):\nwant %v\n got %v", str, pat, want, got)
		}
	}
}

File: python/regex/regex_test.py
import random as rand
import re
from regex import matches_fully

random = rand.Random(0)


def test_random_cases():
	for test in range(50_000):
		str_freq = 1 + random.randrange(20)

		s = ''.join([
			chr(ord('a') + random.randrange(str_freq))
			for _ in range(random.randrange(50))
		])

		pat_chars = ['']*(1+random.randrange(20))
		dot_prob = 1 + random.randrange(20)
		pat_freq = 1 + random.randrange(20)

		j = 0
		while j < len(pat_chars):
			pat_chars[j] = chr(ord('a') + random.randrange(pat_freq))
			if random.randrange(100) < dot_prob:
				pat_chars[j] = '.'

			if j+1 < len(pat_chars):
				prob = random.randrange(100)
				if prob < 20:
					j += 1
					pat_chars[j] = '?'
				elif prob < 40:
					j += 1
					pat_chars[j] = '+'
				elif prob < 60:
					j += 1
					pat_chars[j] = '*'

			j += 1

		pat = ''.join(pat_chars)
		want = re.fullmatch(pat, s) is not None
		got = matches_fully(s, pat)

		assert got == want, f'matches_fully("{s}", "{pat}"):\nwant {want}\n got {got}'
